### DISCUSSION TITLE
What is considered a live trace?

### DISCUSSION THREAD

**Discussion by braunsonm** [grafana/tempo#2380] [Q&A] [answered]
I haven't been able to find any docs on how this works internally in Tempo. We have a large number of apps that are very busy and we recently run up against the 10k default limit for active traces. 
Since traces can be added to with additional spans at anytime, what is an `active` trace? Is it a timer before flushed to disk? This also drives the metric `tempo_ingester_live_traces`

**Comment by mapno** [comment, marked as answer]
Hi! A live trace is a trace that hasn't received a new batch of spans in more than `ingester.trace_idle_period` seconds (default 10s). After that, the trace is flushed to disk and appended to the WAL.

If after that time, Tempo receives a new batch, a new live trace will be created in memory.

**Reply by braunsonm** [reply]
Perfect thank you! I assume if a new batch is received for an existing trace after that, it is merged with the existing trace on disk? (You said a new live trace is created)

**Reply by mapno** [reply]
That's correct. Not immediately, but eventually both batches will get merged into one. Still, during that time, searches for that trace will find all batches.

**Reply by ahurtaud** [reply]
@mapno I think you wanted to mention:
#
A live trace is a trace that **has** received a new batch of spans in **less** than ingester.trace_idle_period seconds (default 10s).

Otherwise I believe all traces become live :)

I am asking this because I see a tenant is throttle due to `max_traces_per_user` limit reach. However I am tryging to find which trace falls into this because whatever I increase it to, it seems one of the trace I receive is always filling this limit ...

**Reply by mapno** [reply]
Yes, absolutely. My definition was for a _dead_ trace 🙈 

You want to know which traces are being kept alive?

**Reply by ahurtaud** [reply]
well basically my tenant (receiving traces from different part of the company I am not much aware of) is constantly reaching max live trace (whatver if I put `max_traces_per_user` to 50k, 200k, or somehow 1M - yes we do have a lot of traces)
However the amount of discarded spans (for that reason) remains stable:
![Screenshot 2025-06-25 at 15 30 34](https://github.com/user-attachments/assets/ecb42a4c-aab7-4b0b-83a4-65aa11e1850a)

left hand side is number of discarded spans with reason : live_trace_exceeded
right hand side is the number of ingesters "current live traces" `tempo_ingester_live_traces`

I ve tried to increase to 200K but the same amount of spans were discarded and my ingestors were somehow going out of memory.

So right now i d like to understand what are those spans, as they may be super weird, like iI dont know infinite traces ...

**Reply by mapno** [reply]
Knowing the topology of the traces they're sending would help (ie. what they look like). If I had to guess, the traces are very small (a few spans each) and the limit is still too low for that—ie. 200k live traces limit for 200k spans/s is not that much if each trace is very small.

**Reply by flu0608** [reply]
Hello,
Is there another definition of "live" trace?
Because i can see that this term is used in the Tempo backend metrics but the meaning seems to me different from the definition you gave.
Here is the Tempo metric:

    # HELP tempodb_backend_objects_total Total number of objects (traces) in the backend
    # TYPE tempodb_backend_objects_total gauge
    tempodb_backend_objects_total{status="compacted",tenant="single-tenant"} 7.6197192e+07
    tempodb_backend_objects_total{status="live",tenant="single-tenant"} 1.9833903679e+10

What i understand from the metric is that there are:
    * 19.8 Bil "live" traces           ==> traces stored in non compacted blocks (in S3 in my case)
    * 76 Mil "compacted" traces ==> traces stored in compacted blocks (in S3 in my case)

Am i wrong?

**Reply by joe-elliott** [reply]
The difference between compacted and live traces in these metrics is discussed here:

https://github.com/grafana/tempo/discussions/5580#discussioncomment-14233799
